src/main/java/org/opensearch/ad/ml/CheckpointDao.java

```diff
@@ -337,10 +349,68 @@ public String toCheckpoint(EntityModel model) {
             if (model.getThreshold() != null) {
                 json.addProperty(ENTITY_THRESHOLD, gson.toJson(model.getThreshold()));
             }
+            if (model.getErcf().isPresent()) {
+                json.addProperty(ENTITY_ERCF, toCheckpoint(model.getErcf().get()));
+            }
             return gson.toJson(json);
         });
     }
 
+    private String toCheckpoint(ExtendedRandomCutForest ercf) {
+        String checkpoint = null;
+        Map.Entry<LinkedBuffer, Boolean> result = checkoutOrNewBuffer();
+        LinkedBuffer buffer = result.getKey();
+        boolean isCheckout = result.getValue();
+        try {
+            checkpoint = toCheckpoint(ercf, buffer);
+        } catch (RuntimeException e) {
+            logger.error("Failed to serialize model", e);
+            if (isCheckout) {
+                try {
+                    serializeRCFBufferPool.invalidateObject(buffer);
+                } catch (Exception x) {
+                    logger.warn("Failed to invalidate buffer", x);
+                }
+                checkpoint = toCheckpoint(ercf, LinkedBuffer.allocate(serializeRCFBufferSize));
+            }
+        } finally {
+            if (isCheckout) {
+                try {
+                    serializeRCFBufferPool.returnObject(buffer);
+                } catch (Exception e) {
+                    logger.warn("Failed to return buffer to pool", e);
+                }
+            }
+        }
```

Comment:
If a `RuntimeException` triggers the `catch` block, the `finally` block will still run afterwards. Is it the correct behavior to still invoke `serializeRCFBufferPool.returnObject(buffer)` after already invoking `serializeRCFBufferPool.invalidateObject(buffer)`?

Reply:
That might not be correct. I changed the behavior to not return it in this case. Thanks for pointing it out!